Optimise array handling of RequestEventDispatcher

gavin-norman-sociomantic opened the pull request.

Commit:
Remove all events for fiber in one go, rather than one by one

At the end of `nextEvent`, all events that were registered for a fiber must
be unregistered. Previously, they were unregistered one by one. This commit
changes the logic to simply unregister all events for the current fiber, in
one go.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -701,6 +681,27 @@ public struct RequestEventDispatcher
+
+    private void unregisterFiber ( MessageFiber fiber )
+    {
+        this.waiting_fibers.length = Array.moveToEnd(this.waiting_fibers,
```
`removeShift` is not suitable here?

gavin-norman-sociomantic replied:
`removeShift` removes a single element, whereas here we want to remove several matching elements in one pass.

gavin-norman-sociomantic replied:
(The 2nd commit refactors this method again, anyway.)

nemanja-boric-sociomantic replied:
> removeShift removes a single element, whereas here we want to remove several matching elements in one pass.

Ah :+1:

gavin-norman-sociomantic commented:
>Refactor RequestEventDispatcher to never modify length of internal array

I suppose this should say "never _reduce_ length", actually.

gavin-norman-sociomantic commented:
Updated: added protection attribute; fixed commit message.

gavin-norman-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -629,39 +634,56 @@ public struct RequestEventDispatcher
+            }
+
+            // Ensure the user isn't trying to do anythign crazy.
+            if ( waiting_fiber.enabled &&
```
Hmm I notice that the main loop to find the matching element can `break`, which means that the sanity check loop doesn't complete. These should probably be two separate loops now.

david-eckardt-sociomantic replied:
Or don't `break` and set `in_list` only if it is not `null`.

gavin-norman-sociomantic replied:
Updated.

gavin-norman-sociomantic replied:
Updated with David's suggestion.

david-eckardt-sociomantic replied:
Is `assert(in_list is null)` still correct now that this loop doesn't `break`?

gavin-norman-sociomantic replied:
Yes, because it should be impossible to register the same fiber/event pair twice.

gavin-norman-sociomantic replied:
(That assertion actually came from an earlier version of the code which didn't break the loop.)

david-eckardt-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -251,9 +251,19 @@ public struct RequestEventDispatcher
+
+        /// Is this instance currently enabled? If false, it will be ignored
+        /// when dispatching events.
+        bool enabled;
```
Could you use `event.active.none` (i.e. `!event.active`) to indicate that this instance is disabled?

gavin-norman-sociomantic replied:
Unfortunately not, as the loops in `register` and `unregister` match the fiber/event pair.

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -629,39 +634,56 @@ public struct RequestEventDispatcher
+                with ( event.Active ) switch ( event.active )
+                {
+                    case message:
                         enforce(waiting_fiber.event.message.type
```
This is something else, not related to this commit, but shouldn't this be `verify` ?

gavin-norman-sociomantic replied:
I suppose you're right. It would be a programmer error, not an error based on processing run-time data. `verify` doesn't exist in the ocean version that swarm uses, but these could be turned into `assert`s.

nemanja-boric-sociomantic replied:
Maybe just add a comment here, so we know when we `verify` everything?

nemanja-boric-sociomantic replied:
Ping here, and LGTM.

nemanja-boric-sociomantic replied:
Two things: 

- `enforce`` <- two ticks at the end
- maybe add `should be asserts (and ideally verify)` so we can `grep` for `assert` when do the transition.

gavin-norman-sociomantic replied:
Ah yes. Updated.

gavin-norman-sociomantic replied:
Oh yeah. Updated. (The comments already contain the string "assert".)

nemanja-boric-sociomantic replied:
Yes, I just saw it, sorry!

nemanja-boric-sociomantic commented on src/swarm/neo/request/RequestEventDispatcher.d:
```diff
@@ -756,11 +783,12 @@ public struct RequestEventDispatcher
+            if ( waiting_fiber.enabled &&
+                waiting_fiber.event.active == waiting_fiber.event.active.send )
             {
                 WaitingFiber ret = waiting_fiber;
```
```
this.disable(waiting_fiber);
return waiting_fiber;
```

or am I missing something?

gavin-norman-sociomantic replied:
I think the idea is to make a copy of it before it's disabled, then return the copy.

gavin-norman-sociomantic replied:
(I don't remember what difference this could make, though.)

nemanja-boric-sociomantic replied:
Ah, could it be due the fact that the old code was moving this instance around the array, invalidating this reference to it? In that case, it's not needed any more.

gavin-norman-sociomantic replied:
Maybe. I'll have a look at the usage.

gavin-norman-sociomantic replied:
Yep, you're right. Looks like a refactoring omission. Updated.

Commit:
Refactor RequestEventDispatcher to never reduce length of internal array

Profiling showed that the array removal operations were taking a signifcant
number of CPU cycles. This commit slightly refactors the struct to mark
each array element as enabled or disabled, taking advantage of the fact
that most requests will continually un/register the same fiber/event pairs
over and over.

Commit:
Add note on adapting enforce -> verify
